File: backend/app/api/terminal.py
import asyncio
import fcntl
import logging
import os
import pty
import select
import signal
import struct
import termios
import time

from fastapi import APIRouter, WebSocket, WebSocketDisconnect, Query
from app.core.auth import verify_ws_token

logger = logging.getLogger("tiup_visualizer")

router = APIRouter()

# Idle timeout in seconds: if no data flows in either direction for this long,
# the server closes the PTY and WebSocket automatically.
IDLE_TIMEOUT = 600  # 10 minutes


@router.websocket("/ws/terminal")
async def terminal_websocket(websocket: WebSocket, token: str = Query(default=None)):
    """WebSocket endpoint that provides an interactive bash terminal via PTY."""
    # Verify authentication token
    username = verify_ws_token(token)
    if username is None:
        await websocket.close(code=4001, reason="Authentication required")
        return

    await websocket.accept()
    logger.info(f"WebSocket terminal opened by user: {username}")

    # Create a pseudo-terminal
    master_fd, slave_fd = pty.openpty()

    # Fork a child process running bash
    pid = os.fork()
    if pid == 0:
        # Child process
        os.close(master_fd)
        os.setsid()

        # Set the slave as the controlling terminal
        fcntl.ioctl(slave_fd, termios.TIOCSCTTY, 0)

        # Redirect stdio to the slave PTY
        os.dup2(slave_fd, 0)
        os.dup2(slave_fd, 1)
        os.dup2(slave_fd, 2)
        if slave_fd > 2:
            os.close(slave_fd)

        # Set environment variables for a nicer terminal experience
        env = os.environ.copy()
        env["TERM"] = "xterm-256color"
        env["COLORTERM"] = "truecolor"

        os.execvpe("/bin/bash", ["/bin/bash", "--login"], env)
    else:
        # Parent process
        os.close(slave_fd)

        # Set master_fd to non-blocking
        flag = fcntl.fcntl(master_fd, fcntl.F_GETFL)
        fcntl.fcntl(master_fd, fcntl.F_SETFL, flag | os.O_NONBLOCK)

        # Set default terminal size
        _set_terminal_size(master_fd, 24, 80)

        # Track whether cleanup has been done to avoid double-cleanup
        cleaned_up = False
        # Shared mutable timestamp for idle detection
        last_activity = time.monotonic()

        def touch_activity():
            nonlocal last_activity
            last_activity = time.monotonic()

        def cleanup():
            nonlocal cleaned_up
            if cleaned_up:
                return
            cleaned_up = True

            # Close master fd
            try:
                os.close(master_fd)
            except OSError:
                pass

            # Terminate child process gracefully, then force-kill if needed
            try:
                os.kill(pid, signal.SIGTERM)
            except OSError:
                return
            # Non-blocking wait with timeout — avoid hanging forever
            for _ in range(10):
                try:
                    result = os.waitpid(pid, os.WNOHANG)
                    if result[0] != 0:
                        return
                except (OSError, ChildProcessError):
                    return
                time.sleep(0.1)
            # Force kill if SIGTERM didn't work within 1 second
            try:
                os.kill(pid, signal.SIGKILL)
                os.waitpid(pid, 0)
            except (OSError, ChildProcessError):
                pass

        async def read_from_pty():
            """Read output from PTY and send to WebSocket."""
            try:
                while True:
                    await asyncio.sleep(0.01)
                    if cleaned_up:
                        break
                    try:
                        r, _, _ = select.select([master_fd], [], [], 0)
                        if r:
                            data = os.read(master_fd, 4096)
                            if not data:
                                # PTY closed (child exited)
                                break
                            touch_activity()
                            await websocket.send_bytes(data)
                    except OSError:
                        break
            except (WebSocketDisconnect, Exception):
                pass

        async def idle_watchdog():
            """Periodically check for idle timeout and force-close if exceeded."""
            try:
                while not cleaned_up:
                    await asyncio.sleep(30)  # check every 30s
                    if cleaned_up:
                        break
                    elapsed = time.monotonic() - last_activity
                    if elapsed >= IDLE_TIMEOUT:
                        logger.info(
                            f"Terminal idle timeout ({IDLE_TIMEOUT}s) for user: {username}, closing"
                        )
                        # Notify client before closing
                        try:
                            await websocket.send_text(
                                "\r\n\x1b[33m[Session timed out due to inactivity]\x1b[0m\r\n"
                            )
                        except Exception:
                            pass
                        # Close websocket to trigger the finally cleanup
                        try:
                            await websocket.close(code=4002, reason="Idle timeout")
                        except Exception:
                            pass
                        break
            except (asyncio.CancelledError, Exception):
                pass

        read_task = asyncio.ensure_future(read_from_pty())
        idle_task = asyncio.ensure_future(idle_watchdog())

        try:
            while True:
                message = await websocket.receive()
                if message.get("type") == "websocket.disconnect":
                    break

                touch_activity()

                if "text" in message:
                    text = message["text"]
                    # Handle resize command
                    if text.startswith("\x1bresize:"):
                        parts = text.split(":")
                        if len(parts) == 3:
                            try:
                                rows = int(parts[1])
                                cols = int(parts[2])
                                _set_terminal_size(master_fd, rows, cols)
                            except (ValueError, OSError):
                                pass
                        continue
                    try:
                        os.write(master_fd, text.encode("utf-8"))
                    except OSError:
                        break
                elif "bytes" in message:
                    try:
                        os.write(master_fd, message["bytes"])
                    except OSError:
                        break
        except WebSocketDisconnect:
            pass
        except Exception:
            pass
        finally:
            read_task.cancel()
            idle_task.cancel()
            try:
                await read_task
            except (asyncio.CancelledError, Exception):
                pass
            try:
                await idle_task
            except (asyncio.CancelledError, Exception):
                pass
            cleanup()
            logger.info(f"WebSocket terminal closed for user: {username}")
            # Close websocket if still open
            try:
                await websocket.close()
            except Exception:
                pass


def _set_terminal_size(fd: int, rows: int, cols: int):
    """Set the terminal size of a PTY."""
    size = struct.pack("HHHH", rows, cols, 0, 0)
    fcntl.ioctl(fd, termios.TIOCSWINSZ, size)

File: backend-go/terminal.go
package main

import (
	"fmt"
	"log/slog"
	"net/http"
	"os"
	"os/exec"
	"strings"
	"sync"
	"syscall"
	"time"
	"unsafe"

	"github.com/creack/pty"
	"github.com/gorilla/websocket"
)

const idleTimeout = 600 * time.Second // 10 minutes

var upgrader = websocket.Upgrader{
	CheckOrigin: func(r *http.Request) bool { return true },
}

func (s *Server) handleTerminal(w http.ResponseWriter, r *http.Request) {
	tokenStr := r.URL.Query().Get("token")
	username, ok := s.auth.VerifyToken(tokenStr)
	if !ok {
		slog.Warn("Terminal auth failed", "remote", r.RemoteAddr)
		http.Error(w, "Authentication required", http.StatusUnauthorized)
		return
	}

	conn, err := upgrader.Upgrade(w, r, nil)
	if err != nil {
		slog.Error("WebSocket upgrade failed", "error", err, "user", username, "remote", r.RemoteAddr)
		return
	}
	defer conn.Close()

	slog.Info("WebSocket terminal opened", "user", username, "remote", r.RemoteAddr)

	// Start bash with PTY
	cmd := exec.Command("/bin/bash", "--login")
	cmd.Env = append(os.Environ(), "TERM=xterm-256color", "COLORTERM=truecolor")

	ptmx, err := pty.Start(cmd)
	if err != nil {
		slog.Error("Failed to start PTY", "error", err, "user", username)
		conn.WriteMessage(websocket.CloseMessage,
			websocket.FormatCloseMessage(websocket.CloseInternalServerErr, "Failed to start terminal"))
		return
	}
	defer ptmx.Close()

	slog.Info("PTY started", "user", username, "pid", cmd.Process.Pid)

	// Set default terminal size
	setTerminalSize(ptmx, 24, 80)

	var (
		cleanedUp    bool
		cleanupMu    sync.Mutex
		lastActivity = time.Now()
		activityMu   sync.Mutex
	)

	touchActivity := func() {
		activityMu.Lock()
		lastActivity = time.Now()
		activityMu.Unlock()
	}

	cleanup := func() {
		cleanupMu.Lock()
		defer cleanupMu.Unlock()
		if cleanedUp {
			slog.Debug("Cleanup already done", "user", username)
			return
		}
		cleanedUp = true

		slog.Info("Cleaning up terminal", "user", username, "pid", cmd.Process.Pid)

		ptmx.Close()

		if cmd.Process != nil {
			cmd.Process.Signal(syscall.SIGTERM)
			done := make(chan struct{})
			go func() {
				err := cmd.Wait()
				slog.Info("Process exited", "user", username, "pid", cmd.Process.Pid, "wait_err", err)
				close(done)
			}()
			select {
			case <-done:
				slog.Debug("Process exited gracefully", "user", username, "pid", cmd.Process.Pid)
			case <-time.After(1 * time.Second):
				slog.Warn("Force killing process", "user", username, "pid", cmd.Process.Pid)
				cmd.Process.Signal(syscall.SIGKILL)
				cmd.Wait()
			}
		}
	}
	defer cleanup()

	// Read from PTY, send to WebSocket
	done := make(chan struct{})
	go func() {
		defer close(done)
		buf := make([]byte, 4096)
		for {
			n, err := ptmx.Read(buf)
			if err != nil {
				slog.Info("PTY read ended", "user", username, "error", err)
				// PTY closed (bash exited) - close WebSocket to notify frontend
				conn.WriteMessage(websocket.CloseMessage,
					websocket.FormatCloseMessage(websocket.CloseNormalClosure, "Process exited"))
				return
			}
			if n > 0 {
				touchActivity()
				if err := conn.WriteMessage(websocket.BinaryMessage, buf[:n]); err != nil {
					slog.Info("WebSocket write failed", "user", username, "error", err)
					return
				}
			}
		}
	}()

	// Idle watchdog
	go func() {
		ticker := time.NewTicker(30 * time.Second)
		defer ticker.Stop()
		for {
			select {
			case <-done:
				return
			case <-ticker.C:
				activityMu.Lock()
				elapsed := time.Since(lastActivity)
				activityMu.Unlock()

				if elapsed >= idleTimeout {
					slog.Info("Terminal idle timeout", "user", username, "timeout", idleTimeout)
					conn.WriteMessage(websocket.TextMessage,
						[]byte("\r\n\x1b[33m[Session timed out due to inactivity]\x1b[0m\r\n"))
					conn.WriteMessage(websocket.CloseMessage,
						websocket.FormatCloseMessage(4002, "Idle timeout"))
					return
				}
			}
		}
	}()

	// Read from WebSocket, write to PTY
	for {
		msgType, msg, err := conn.ReadMessage()
		if err != nil {
			slog.Info("WebSocket read ended", "user", username, "error", err, "msgType", msgType)
			break
		}

		touchActivity()

		switch msgType {
		case websocket.TextMessage:
			text := string(msg)
			// Handle resize command: \x1bresize:rows:cols
			if strings.HasPrefix(text, "\x1bresize:") {
				parts := strings.Split(text, ":")
				if len(parts) == 3 {
					var rows, cols int
					if _, err := fmt.Sscanf(parts[1]+":"+parts[2], "%d:%d", &rows, &cols); err == nil {
						setTerminalSize(ptmx, rows, cols)
						slog.Debug("Terminal resized", "user", username, "rows", rows, "cols", cols)
					}
				}
				continue
			}
			if _, err := ptmx.Write(msg); err != nil {
				slog.Info("PTY write failed", "user", username, "error", err)
				goto exit
			}
		case websocket.BinaryMessage:
			if _, err := ptmx.Write(msg); err != nil {
				slog.Info("PTY write failed", "user", username, "error", err)
				goto exit
			}
		}
	}

exit:
	slog.Info("WebSocket terminal closed", "user", username, "pid", cmd.Process.Pid)
}

func setTerminalSize(f *os.File, rows, cols int) {
	ws := struct {
		Row    uint16
		Col    uint16
		Xpixel uint16
		Ypixel uint16
	}{
		Row: uint16(rows),
		Col: uint16(cols),
	}
	syscall.Syscall(
		syscall.SYS_IOCTL,
		f.Fd(),
		uintptr(syscall.TIOCSWINSZ),
		uintptr(unsafe.Pointer(&ws)),
	)
}
